### PR TITLE
status reporting

### PR DESCRIPTION
**Pull request by mwear**
**Description:**
I started a POC for status reporting in https://github.com/open-telemetry/opentelemetry-collector/pull/5158. That PR served as a good starting point for discussion and feedback, but the implementation has changed direction enough, that I decided to close it in favor of this PR. 

As a brief recap, this work was spun out of a discussion in the contrib repo around improving startup behavior, and error reporting at startup. See:

* https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/8816
* https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/8817

A proposal was suggested in a contrib issue: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/8816#issuecomment-1083380460

> The proposal is then to change the component.Host to allow components to report its state (and continue reporting it), adding another function allowing components to register for callbacks, to be called whenever there's a report. This way, components such as the load balancing exporter and the health check extension can register for callbacks, whereas scrapers and components like the OTLP exporter can report their connection failures in real time.

This PR attempts to provide a status reporting mechanism as described in the above comment. To that end, the `Host` interface was extended with a `ReportStatus` function that components can use to report `ok`, or `error` statuses, and a `RegisterStatusListener` function, where interested components (i.e. the healthcheck extension, etc) can register callbacks to receive status notifications.

In discussions on #5158 and at the SIG meeting, it was suggested that the same mechanism could replace the `PipelineWatcher` interface. `RegisterStatusListener` also allows optional registration of `PipelineReady` and `PipelineNotReady` callbacks. I have not removed the PipelineWatcher interface in this PR, but can (and will) remove it in a follow up PR when (and if) we decide this is the direction we want to head in.

Another suggestion was that this system could also replace the `Host.ReportFatalError` function. If `Host.ReportStatus` is called with a `status.Event` of type `status.FatalError` it will behave the same way as `Host.ReportFatalError` does today.

Another requirement is that this mechanism should work with both full and partial restarts of the service. The idea is that components will register status listeners during startup, and unregister during shutdown. So long as this convention is followed, we should always have active components reporting and listening for status notifications. We might need to discuss and adjust what pipeline ready, or not ready means in the context of a partial restart, but I think we can come up with a reasonable way to accommodate that scenario.

~Lastly, we probably should discuss the new error types introduced in this PR. This PR adds `recoverable`, `permanent`, and `fatal` to the consumererror package. The use cases for `recoverable` and `fatal` are pretty self explanatory. A fatal error will cause the collector to shutdown, whereas a recoverable error, is likely transient and could resolve over time. The permanent case, is somewhat less obvious, and it should be noted that the `consumererror` package has the same error type. If we keep it, we should try to find a common place for the error to reside. I could see a permanent error resulting from a component retrying recoverable errors for a time period. Once it has exhausted the retry strategy, it might choose to shut itself down and return a permanent error to indicate early shutdown of a component. I'm not sure if we actually have this use case today, or whether it's imaginary at this point.~

**Basic usage:**

**Register and Unregister**

```golang
// All callbacks are optional. A listener can register as many, or few as desired. Some listeners
// might care about status events only, other might only be interested in pipeline ready and / or
// not ready.
func (c *mycomponent) start(_ context.Context, host component.host) {
    c.unregisterFunc := host.RegisterStatusListener(
        component.WithStatusEventHandler(statusEventHandler),
        component.WithPipelineReadyHandler(readyHandler),
        component.WithPipelineNotReadyHandler(notReadyHandler),
    )
}

func (c *mycomponent) statusEventHandler(event component.StatusEvent) {
    switch event.Type {
        case component.StatusOK:
		// component is ok (or has recovered)
	case component.StatusRecoverableError:
		//handle recoverable
        case component.StatusPermanentError:
		//handle permanent
        case component.StatusFatalError:
		//handle fatal, collector will shutdown after listener callbacks are executed
	default:
		//handle unset
    }
}

func (c *mycomponent) shutdown(_ context.Context) {
    // call the unregisterFunc during shutdown, or when appropriate
    err := c.unregisterFunc()
}

```

**ReportStatus**

```golang
// recoverable error
if err := s.connect(); err != nil {
    ev, _ := componet.NewStatusEvent(component.StatusRecoverableError, component.WithError(err))
    host.ReportStatus(ev)
}

// permanent error
if err := s.connect(); err != nil {
    ev, _ := component.NewStatusEvent(component.StatusPermanentError, component.WithError(err))
    host.ReportStatus(ev)
}

// fatal error
if err := s.superImportant(); err != nil {
    ev, _ := component.NewStatusEvent(component.FatalError, component.WithError(err))
    host.ReportStatus(ev)
}

// everything is fine
if err := s.connect(); err == nil {
    ev, _ := component.NewStatusEvent(component.StatusOK)
    host.ReportStatus(ev)
}

// status.OK and status.WithError not valid
if ev, err := component.NewStatusEvent(component.StatusOK, component.WithError(err)); err != nil {
    logger.Warn("error provided with ok status")
}
```

**Summary**

To summarize the above, this PR:
- introduces a status reporting mechanism for components to report status, and interested components to register callbacks to listen for status
- provides a potential replacement for the `PipelineWatcher` interface
- provides a potential replacement for `Host.ReportFatalError` function
- ~introduces a `componenterror` package with `fatal`, `recoverable`, and `permanent` errors~

**Link to tracking Issue:** 
See:
- https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/8816#issuecomment-1083380460
- https://github.com/open-telemetry/opentelemetry-collector/pull/5158

**Testing:**
- Mainly unit testing

**Documentation:**
- Source code documentation

### CODE REVIEW

File: component/host.go [hunk 1]
```diff
@@ -63,4 +64,14 @@ type Host interface {
 	// GetExporters can be called by the component anytime after Component.Start() begins and
 	// until Component.Shutdown() ends.
 	GetExporters() map[config.DataType]map[config.ComponentID]Exporter
+
+	// ReportStatus can be used by a component to communicate its status to registered status
+	// listeners. Components can use this function to indicate that they are functioning properly,
+	// or are in an error state.
+	ReportStatus(eventType status.EventType, componentID config.ComponentID, options ...status.StatusEventOption)
```
Comment (mx-psi) [inline component/host.go:71]: AFAICT, the `componentID` should be the same on each call that a given component does. Should we therefore remove the `componentID` argument from the public interface? Instead, we could wrap the host we pass to each component so that it sets this automatically.

If we ever need to have a component 'lie' about its component ID, this could still be achieved via an Option.

File: component/status/status.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+)
+
+// EventType represents an enumeration of status event types
+type EventType int
+
+const (
+	// EventOK indicates the producer of a status event is functioning normally
+	EventOK EventType = iota
+	// EventError indicates the producer of a status event is in an error state
+	EventError
```
Comment (bogdandrutu) [inline component/status/status.go:30]: Why not to extend this list with: OK, RECOVERABLE_ERROR, PERMANENT_ERROR, FATAL_ERROR **vs** use the combination of EventError + componenterror type?

File: component/status/notifications_test.go [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+
+	un := notifications.RegisterListener(
+		WithPipelineReadyHandler(l.PipelineReadySpy.Func),
+		WithPipelineNotReadyHandler(l.PipelineNotReadySpy.Func),
+	)
+
+	assert.False(t, l.PipelineReadySpy.WasCalled(), "Unexpected call to PipelineReady")
+	assert.False(t, l.PipelineReadySpy.WasCalled(), "Unexpected call to PipelineNotReady")
+
+	notifications.Start()
```
Comment (bogdandrutu) [inline component/status/notifications_test.go:37]: assert.NoError(t, notifications.Start())

File: component/status/notifications.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+)
+
+var errRegistrationNotFound = errors.New("registration not found")
+
+// Notifications is a mechanism that facilitates the passing of status events between components.
+// Status events can be used to indicate that a component is functioning as expected, or that it is
+// in an error state. Components can send status events using the ReportStatus function, and
+// components interested in receiving events can register a listener via the RegisterListener
+// function.
+type Notifications struct {
```
Comment (bogdandrutu) [inline component/status/notifications.go:32]: why does this need to be public? Can it be in an internal package for the moment?

File: component/host.go (hunk 1 above)
Comment (mwear) [reply component/host.go]: I added the functionality to `hostWrapper`. I had to plumb through the component id by attaching it to the built components. Let me know if there is a better way.

File: component/status/status.go (hunk 2 above)
Comment (mwear) [reply component/status/status.go]: I made this change and got rid of the newly introduced error types.

File: component/status/status.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+	return func(o *StatusEvent) {
+		o.Error = err
+	}
+}
+
+// WithTimestamp sets the timestamp, expected in nanos, for a StatusEvent
+func WithTimestamp(nanos int64) StatusEventOption {
+	return func(o *StatusEvent) {
+		o.Timestamp = nanos
+	}
```
[the remark this answers is not in the document]
Comment (mwear) [reply component/status/status.go]: I got rid of the options and left everything mutable. I would have gone the immutable route, but we need to be able to set the component id via the `hostWrapper`. I also got rid of `NewStatusEvent`. It was useful for initializing the timestamp. I moved that to the `hostWrapper` for now, but it isn't obvious that it gets set there unless you thoroughly study the code. Should I have functions to help construct a StatusEvent even with all fields exported? Would `NewStatusEvent(eventType status.EventType)` and `NewStatusEventWithError(eventType status.EventType, err error)` be helpful or something else?

File: component/status/notifications.go (hunk 4 above)
Comment (mwear) [reply component/status/notifications.go:32]: It's public so it can be used in the service package. I think I could move it to the service if I made status.Listener public. Would that move make sense?

Comment (bogdandrutu) [reply component/status/notifications.go:32]: you can also move to a directory that includes "internal" in the path (see https://go.dev/doc/go1.4#internalpackages)

Comment (mwear) [reply component/status/notifications.go]: It is now internal within the service subtree.

File: component/status/status.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (mx-psi) [reply component/status/status.go]: >  I would have gone the immutable route, but we need to be able to set the component id via the hostWrapper

A way around this is to define the status on `service/internal/components` (or another suitable internal package where you also define the host wrapper) and then define an alias for it (`type StatusEvent = component.StatusEvent`) on `component/status`.

[the remark this answers is not in the document]
Comment (mwear) [reply component/status/status.go]: I generally like this idea and I see how this could allow us to update the component id from the host wrapper, but I sketched out the details of what I think this would require, and I'm not sure we want to go this direction. Since the host wrapper is in `service/internal/components` we would have to define `StatusEvent` in the same package, and alias it in the `service` package to make it publicly available as `service.StatusEvent`. If we wanted to make it available from the component package we'd have to alias it a second time to have `component.StatusEvent`. We would have two aliases, which seems less than ideal. Since the goal would be to make the `StatusEvent` immutable to the outside world, we'd have to do the same for the constructor and the `StatusEventOption`s as well. One last issue, is that I tried to alias a type from `service/internal` to `service` to `component` and found that it creates a circular dependency (e.g. import cycle not allowed). It's possible I'm going about this the wrong way, if so, let me know, but if this is accurate, it seems like this path has some unintended consequences.

As things are currently implemented the `service.ReportStatus` function has a chance to modify a `StatusEvent` before passing it off to listeners (to update component id, for example). The event is passed by value to the listeners, so a change to a field by one listener, will not be seen by other listeners. Given that, I'm not overly concerned with all the fields being public on `StatusEvent`. If others have concerns, let me know. I wish we had a better way to initialize the timestamp when creating an event though.

[the remark this answers is not in the document]
Comment (mwear) [reply component/status/status.go]: After thinking this through, I think we could make our lives easier by making `StatusEvent` immutable and update it by making a copy with the new component id in the host wrapper. I'll move forward with that approach if there aren't any objections.

File: component/host.go (hunk 1 above)
Comment (mx-psi) [reply component/host.go]: This is roughly what I had in mind!

File: component/host.go [hunk 6]
```diff
@@ -63,4 +64,14 @@ type Host interface {
 	// GetExporters can be called by the component anytime after Component.Start() begins and
 	// until Component.Shutdown() ends.
 	GetExporters() map[config.DataType]map[config.ComponentID]Exporter
+
+	// ReportStatus can be used by a component to communicate its status to registered status
+	// listeners. Components can use this function to indicate that they are functioning properly,
+	// or are in an error state.
+	ReportStatus(event *status.Event)
```
[the remark this answers is not in the document]
Comment (mwear) [reply component/host.go:72]: Noting that we would also like to replace the PipelineWatcher interface with the functionality provided in this PR, but in order to keep the diff manageable, I anticipate addressing it in a followup PR.

File: CHANGELOG.md [hunk 7]
```diff
@@ -43,6 +43,7 @@
 
 - Extend config.Map.Unmarshal hook to check map key string to any TextUnmarshaler not only ComponentID (#5244)
 - Collector will no longer print error with stack trace when the collector is shutdown due to a context cancel. (#5258)
+- Adds status reporting, which allows status events to be sent between components. It also provides functionality to replace the PipelineWatcher interface by registering callbacks for pipeline events, and functionality to replace `Host.ReportFatalError`, by sending a status event via `Host.ReportStatus` of type `status.FatalError` (#5304)
```
[the remark this answers is not in the document]
Comment (mwear) [reply CHANGELOG.md]: This PR has an enhancement, a breaking change, and a deprecation. I went ahead a put a note under each of those headings. If that is TMI, let me know and I can condense it under the breaking section.

[the remark this answers is not in the document]
Comment (mx-psi) [reply CHANGELOG.md]: The more information the merrier IMO :) Thanks!

**Review by mx-psi — COMMENTED** [review]
Left a couple more comments. Other than that, the PR looks good to me. 

The only remaining unsatisfactory thing is that we allow changing the events' component ID and timestamp with no clear use case in mind. IIUC we do that to allow the host wrapper to set these fields, but I wish we did this in a way that didn't also allow components to modify these fields. Since I can't think of a good way to solve this and we can still break things in the future, I am fine with leaving this problem for later and seeing how this API fares with existing components.

File: component/status/status.go [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+
+// WithTimestamp sets the timestamp, expected in nanos, for a status.Event
+func WithTimestamp(nanos int64) EventOption {
+	return func(o *Event) {
+		o.timestamp = nanos
+	}
+}
+
+// WithError assigns an error object to a Event. It is optional and should only be applied
+// to a Event with type status.EventError.
```
[the remark this answers is not in the document]
Comment (mwear) [reply component/status/status.go]: I updated `EventOption` (and the comment, which was slightly outdated). An error is valid with `status.RecoverableError`, `status.PermanentError`, and `status.FatalError`  (everything but `status.OK`) and this is now enforced during construction.

File: component/status/status.go [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+	PermanentError
+	// FatalError is an error that cannot be recovered from and will cause early termination of the collector
+	FatalError
+)
+
+// Event is a status event produced by a component to communicate its status to registered listeners.
+// An event can signal that a component is working normally (i.e. Type: status.OK), or that it
+// is in an error state (i.e. Type: status.RecoverableError). An error status may optionally
+// include an error object to provide additional insight to registered listeners.
+type Event struct {
```
Comment (bogdandrutu) [inline component/status/status.go:42]: I think we also need to know if this is a "receiver/processor/exporter"

File: component/status/status.go [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+
+	"go.opentelemetry.io/collector/config"
+)
+
+// EventType represents an enumeration of status event types
+type EventType int
+
+const (
+	// OK indicates the producer of a status event is functioning normally
+	OK EventType = iota
```
[the remark this answers is not in the document]
Comment (mwear) [reply component/status/status.go:29]: On the prototype @pmm-sumo pointed me towards this document: https://docs.google.com/document/d/15nJTAzomzUJvwhINvTxHuaxUNe7dfgfG3gS1Tib8TGM/edit#. This screenshot is taken from it:
Screen Shot 2022-05-16 at 5 54 02 PM

It shows a component reporting a stream of status events. I think the conventions need to be better defined, and I could use some help from the author of that document (if anyone knows who it is) to solidify them. My initial thought is that a component should report its status on startup, and if they do work on a periodic basis, they should report the result of that work (an export result for an exporter, or the result of a scrape for a scraping receiver, etc). I'm open to suggestions on all of this.

File: service/internal/builder/pipelines_builder.go [hunk 11]
```diff
@@ -52,7 +53,7 @@ type BuiltPipelines map[config.ComponentID]*builtPipeline
 func (bps BuiltPipelines) StartProcessors(ctx context.Context, host component.Host) error {
 	for _, bp := range bps {
 		bp.logger.Info("Pipeline is starting...")
-		hostWrapper := components.NewHostWrapper(host, bp.logger)
+		hostWrapper := components.NewHostWrapper(host, bp.componentID, bp.logger)
```
[the remark this answers is not in the document]
Comment (mwear) [reply service/internal/builder/pipelines_builder.go:56]: I see. I will have to have host wrapper per processor then, and find a way to easily access the component id. Speaking of which, is there a reason why the component id is not part of the component interface?

File: component/status/status.go [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+// UnregisterFunc is a function to be called to unregister a component that has previously
+// registered to listen to status notifications
+type UnregisterFunc func() error
+
+var noopStatusEventFunc = func(event *Event) error { return nil }
+
+var noopPipelineFunc = func() error { return nil }
+
+// Listener is a struct that manages handlers to status and pipeline events
+type Listener struct {
```
[the remark this answers is not in the document]
Comment (mwear) [reply component/status/status.go:135]: This struct is being used by the notifications implementation in service/internal/components. When the notifications implementation was in component I was able to keep it private. The public helpers are so that it is immutable after construction. Is there a better way to organize the package layout to hide it? I was just wondering if you have anything in mind, no problem if you don't, and I can think more about this if it's important.

[the remark this answers is not in the document]
Comment (bogdandrutu) [reply component/status/status.go:135]: But this can also be in a internal package correct?

[the remark this answers is not in the document]
Comment (bogdandrutu) [reply component/status/status.go:135]: For me it is very important to not expose anything public that does not need to be public.

File: service/internal/builder/pipelines_builder.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply service/internal/builder/pipelines_builder.go:56]: Yes, it is for the moment used in the config.Receiver interface, so to avoid circular dependencies needs to be there. I have a proposal to change that.

File: component/status/status.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (pmm-sumo) [reply component/status/status.go:29]: @mwear that would be me :) 

The document was describing ideas on better own telemetry reporting, largely with early work on [remote management](https://github.com/open-telemetry/opamp-spec) in mind. 

Health event reporting I mentioned there is somewhat based on the [Health Events](https://help.sumologic.com/Manage/Health_Events#health-event-log-example) capability at Sumo. At a given time, each component can have _healthy_ or _unhealthy_ status which could be reported upstream. To avoid state flapping, there's the concept of having some sort of threshold before the state changes. Having said that, that was just an idea (and it wasn't included in OpAMP spec). I think the most important thing now is to define the internal API for status reporting (which can then be used for generating health events like described in the docs using status listeners and such).

Let me know if you need any more details on that (also feel free to ping me on CNCF Slack to make it more synchronous, my display name there is "Przemek Maciolek")

File: service/internal/builder/pipelines_builder.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (mwear) [reply service/internal/builder/pipelines_builder.go]: I could not find a way to access to the component id of the processor in `StartProcessors`, so I added a map to `builtPipeline` to lookup the component id by processor. I don't love this solution, but I couldn't find anything I liked better.

File: component/status/status.go (hunk 9 above)
Comment (mwear) [reply component/status/status.go]: I added a `component.Kind` field to the event. Previously, the status reporting structs were in a `status` package (in `component/status`), but this forced me to move everything into component since this addition introduced an import cycle. Many things gained a `Status` prefix during this change.

File: component/status/status.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (mwear) [reply component/status/status.go]: There are two packages that depend on the status listener `component` and `service/internal/components`. `host.RegisterStatusListener` takes parameters of type `StatusListenerOption` to used to construct a StatusListener. After construction, the StatusListener is used in `service/internal/components`. If the `StatusListener` was only used in the `components` package, I could see how we could put it in an internal package, but since it's also used in `service/internal/components` this complicates things. I can not see an easy way to move this to an internal package where everything that needs it will have access to it. Let me know if I am missing something or alternatively, if there is a better way to separate the options that are passed to `host.RegisterStatusListener` from the construction of the `StatusListener` as I think that is root cause of this issue.

**Comment by jpkrohling** [comment]
Are there any news about this one?

**Comment by jpkrohling** [comment]
@bogdandrutu, do you see this PR being merged anytime soon? I'd rather close it, as this has been with no activity since March, with my message for a status request on Jul 4 going unanswered.

**Comment by jpkrohling** [comment]
Closing.
